kernel/kernel-api/src/main/java/io/delta/kernel/internal/tablefeatures/TableFeatures.java

```diff
@@ -17,286 +17,616 @@
+  public static final TableFeature TYPE_WIDENING_PREVIEW_TABLE_FEATURE =
+      new TypeWideningTableFeature("typeWidening-preview");
+
+  private static class TypeWideningTableFeature extends TableFeature.ReaderWriterFeature
+      implements FeatureAutoEnabledByMetadata {
+    TypeWideningTableFeature(String featureName) {
+      super(featureName, /* minReaderVersion = */ 3, /* minWriterVersion = */ 7);
+    }
+
+    @Override
+    public boolean metadataRequiresFeatureToBeEnabled(Protocol protocol, Metadata metadata) {
+      return TableConfig.TYPE_WIDENING_ENABLED.fromMetadata(metadata);
+    }
+
+    @Override
+    public boolean hasKernelWriteSupport(Metadata metadata) {
+      return false; // TODO: yet to support it.
+    }
+  }
+
+  public static final TableFeature IN_COMMIT_TIMESTAMP_W_FEATURE =
+      new InCommitTimestampTableFeature();
+
+  private static class InCommitTimestampTableFeature extends TableFeature.WriterFeature
+      implements FeatureAutoEnabledByMetadata {
+    InCommitTimestampTableFeature() {
+      super("inCommitTimestamp", /* minWriterVersion = */ 7);
+    }
+
+    @Override
+    public boolean metadataRequiresFeatureToBeEnabled(Protocol protocol, Metadata metadata) {
+      return TableConfig.IN_COMMIT_TIMESTAMPS_ENABLED.fromMetadata(metadata);
+    }
   }
 
+  public static final TableFeature TIMESTAMP_NTZ_RW_FEATURE = new TimestampNtzTableFeature();
+
+  private static class TimestampNtzTableFeature extends TableFeature.ReaderWriterFeature
+      implements FeatureAutoEnabledByMetadata {
+    TimestampNtzTableFeature() {
+      super("timestampNtz", /* minReaderVersion = */ 3, /* minWriterVersion = */ 7);
+    }
+
+    @Override
+    public boolean hasKernelWriteSupport(Metadata metadata) {
+      return false; // TODO: yet to be implemented in Kernel
+    }
+
+    @Override
+    public boolean metadataRequiresFeatureToBeEnabled(Protocol protocol, Metadata metadata) {
+      return hasTypeColumn(metadata.getSchema(), TIMESTAMP_NTZ);
+    }
+  }
+
+  public static final TableFeature CHECKPOINT_V2_RW_FEATURE = new CheckpointV2TableFeature();
+
   /**
-   * Checks if the table protocol supports the "domainMetadata" writer feature.
-   *
-   * @param protocol the protocol to check
-   * @return true if the "domainMetadata" feature is supported, false otherwise
+   * In order to commit, there is no extra work required when v2 checkpoint is enabled. This affects
+   * the checkpoint format only. When v2 is enabled, writing classic checkpoints is still allowed.
    */
-  public static boolean isDomainMetadataSupported(Protocol protocol) {
-    return isWriterFeatureSupported(protocol, DOMAIN_METADATA_FEATURE_NAME);
+  private static class CheckpointV2TableFeature extends TableFeature.ReaderWriterFeature
+      implements FeatureAutoEnabledByMetadata {
+    CheckpointV2TableFeature() {
+      super("v2Checkpoint", /* minReaderVersion = */ 3, /* minWriterVersion = */ 7);
```

Comment:
Nit: would it make sense to use the constants like `TABLE_FEATURES_MIN_READER_VERSION` etc?